src/protocol/request_response/mod.rs

```diff
@@ -1057,9 +1062,9 @@ impl RequestResponseProtocol {
                 _ = self.pending_outbound_responses.next(), if !self.pending_outbound_responses.is_empty() => {}
 
                 // Inbound requests that are moved to `pending_outbound_responses`.
-                event = self.pending_inbound_requests.next() => match event {
-                    Some(((peer, request_id), message)) => {
-                        if let Err(error) = self.on_inbound_request(peer, request_id, message).await {
+                event = self.pending_inbound_requests.next(), if !self.pending_inbound_requests.is_empty() => match event {
+                    Some((peer, request_id, request, substream)) => {
+                        if let Err(error) = self.on_inbound_request(peer, request_id, request, substream).await {
```

Comment:
Is there a way to check here if `substream` points to alive substream over an existing connection, and not process the request if not? May be we can do this check in `on_inbound_request`.

Reply:
If the connection has been closed, we'll remove the peer-associated data, which will lead to no further polling on the substream:
https://github.com/paritytech/litep2p/blob/4bc31d03a1af5978059e8a571c4276dae5917718/src/protocol/request_response/mod.rs#L302-L307

The inbound request is directly rejected as a first step here:
https://github.com/paritytech/litep2p/blob/4bc31d03a1af5978059e8a571c4276dae5917718/src/protocol/request_response/mod.rs#L477-L489

If the connection has been closed and detected by the substream, we'll return a `Substream::Closed` error which will not propagate the message:
https://github.com/paritytech/litep2p/blob/4bc31d03a1af5978059e8a571c4276dae5917718/src/protocol/request_response/mod.rs#L501-L511

Reply:
i mean if we received a request and hold it in `pending_inbound_requests`, but the peer gets disconnected before we poll `pending_inbound_requests`, we will process the request anyway and only hit the error when try to send the response to a closed substream? Is there an easy way to not process the inbound request (as it may require resources on the client/runtime side etc.) in this case and fail early?

Reply:
Had a chat offline: we'll not process the inbound requests because we remove the peer state in `on_connection_closed`, have added a comment to make the code easier to follow : https://github.com/paritytech/litep2p/pull/321/commits/58d574266c76f880a0ad4244793f1422eea75a18 🙏